![[Pasted image 20241124210724.png]]
```java
public class NetworkServiceV1_3 {  
  
    public void sendMessage(String data) {  
        String address = "http://example.com";  
        NetworkClientV1 client = new NetworkClientV1(address);  
        client.initError(data);  
  
        String connectResult = client.connect();  
        if (isError(connectResult)) {  
            System.out.println("[네트워크 오류 코드] : " + connectResult);  
        } else {  
            String sendResult = client.send(data);  
            if (isError(sendResult)) {  
                System.out.println("[네트워크 오류 코드] : " + sendResult);  
            }  
        }  
        client.disconnect();  
  
    }  
  
    private static boolean isError(String connectResult) {  
        return !connectResult.equals("success");  
    }  
}
```
- 코드를 보면 정상 흐름과 예외 흐름이 전혀 분리되어 있지 않다.
- 어떤 부분이 정상 흐름이고, 어떤 부분이 예외 흐름인지 이해하기 어렵다.
- 심지어 예외 흐름을 처리하는 부분이 더 많다.


>[!Question] 어떻게 하면 정상 흐름과 예외 흐름을 분리할 수 있을까?
>지금과 같이 반환 값을 사용해서 예외 상황을 처리하는 방식으로는 해결할 수 없다.
>이런 문제를 해결하기 위해 예외 처리 매커니즘이 존재한다.

>[!Success] 예외 처리를 사용하면 정상 흐름과 예외 흐름을 명확하게 분리할 수 있다.

# 예외 계층
![[Pasted image 20241124211117.png]]
- Throwable: 최상위 예외
- Error:
	- 메모리 부족이나 심각한 시스템 오류와 같이 애플리케이션에서 복구가 불가능한 시스템 예외이다.
	- 애플리케이션 개발자는 이 예외를 잡으려고 해서는 안된다.
- Exception
	- **체크 예외** : 개발자가 명시적으로 처리해야 한다.**그렇지 않으면 컴파일 에러 발생**
	- 애플리케이션 로직에서 사용할 수 있는 실질적인 최상위 예외
	- Exception과 그 하위 예외는 모두 **컴파일러가 체크하는 체크 예외**이다.
	- 단, RuntimeException은 예외로 한다.
- RuntimeException
	- **언체크 예외, 런타임 예외**: 명시적으로 **처리하지 않아도 된다.**
	- RuntimeException과 그 자식 예외는 모두 언체크 예외이다.

![[Pasted image 20241124212158.png]]

- 예외는 잡아서 처리하거나 밖으로 던져야 한다.
- 예외를 잡거나 던질 때 지정한 예외뿐만 아니라 그 예외의 자식들도 함께 처리할 수 있다.
- 자바 `main()` 밖으로 예외를 던지면 예외 로그를 출력하면서 시스템이 종료된다.

<정상 흐름과 예외 흐름 분리>
```java
public class NetworkServiceV2_5 {  
  
    public void sendMessage(String data) {  
        String address = "http://example.com";  
        NetworkClientV2 client = new NetworkClientV2(address);  
        client.initError(data);  
  
        try {  
            client.connect();  
            client.send(data);  
  
        } catch (NetworkClientExceptionV2 e) {  
            System.out.println("[오류]코드: " + e.getErrorCode() + ", 메시지: " + e.getMessage());  
        } finally {  
            client.disconnect();  
        }    }  
}
```